weave/flow/eval_imperative.py

```diff
@@ -0,0 +1,314 @@
+            # TODO: can't use score here because it will cause version mismatch
+            # return score
+            return cast(ScoreType, current_score.get())
+
+        scorer.__dict__["score"] = MethodType(score_method, scorer)
+
+        # attach the score feedback to the predict call
+        with call_context.set_call_stack(
+            [self.evaluate_call, self.predict_and_score_call]
+        ):
+            with _set_current_score(score):
+                await self.predict_and_score_call.apply_scorer(scorer)
+
+
+class ImperativeEvaluationLogger(BaseModel):
+    """This class provides an imperative interface for logging evaluations.
+
+    An evaluation is started automatically when the first prediction is logged
+    using the `log_prediction` method, and finished when the `log_summary` method
+    is called.
+
+    Each time you log a prediction, you will get back an `ImperativePredictionLogger`
+    object.  You can use this object to log scores and metadata for that specific
+    prediction (see that class for more details).
+
+    Example:
+        ```python
+        ev = ImperativeEvaluationLogger()
+        pred = ev.log_prediction(inputs, output)
+        pred.log_score(scorer_name, score)
+        ev.log_summary(summary)
+        ```
+    """
+
+    model: Annotated[
+        Model | dict | str,
+        BeforeValidator(_cast_to_cls(Model)),
+        Field(
+            default_factory=Model,
+            description="A metadata-only Model used for comparisons."
+            "Alternatively, you can pass a dict of attributes or just a string"
+            "representing the ID of your model.",
+        ),
+    ]
+
+    _eval_started: bool = PrivateAttr(False)
+    _logged_summary: bool = PrivateAttr(False)
+    _evaluate_call: Call | None = PrivateAttr(None)
+    _pseudo_evaluation: Evaluation = PrivateAttr(
+        default_factory=lambda: Evaluation(
+            dataset=weave.Dataset(rows=weave.Table([{"": ""}])),
+            scorers=[],
+        )
+    )
+
+    def log_prediction(self, inputs: dict, output: Any) -> ImperativeScoreLogger:
+        # similar to how we dynamically create the scorer class, we will
+        # dynamically create the model class
+
+        if not self._eval_started:
+            self._eval_started = True
+
+            # The following section is a "hacky" way to create Model and Evaluation
+            # objects that "look right" to our object saving system.
+
+            # --- Setup the model object ---
+            @weave.op
+            def predict(self: Model, inputs: dict) -> Any:
+                # Get the output from the context variable
+                return current_output.get()
+
+            self.model.__dict__["predict"] = MethodType(predict, self.model)
+
+            # --- Setup the evaluation object ---
+            @weave.op(name="Evaluation.evaluate")
+            def evaluate(self: Evaluation, model: Model) -> None: ...
+
+            @weave.op
```

Comment:
can we call this (`name=Evaluation.predict_and_score`) to match the current datamodel